core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java

```diff
@@ -544,7 +587,105 @@ public GroupByIdBlock getResult()
+        else {
+            batchedPutIfAbsentNoNull(block, blockOffset, batchSize, batchGroupIds, batchGroupIdOffset);
+        }
+    }
+
+    private void batchedPutIfAbsentNullable(Block block, int blockOffset, int batchSize, long[] batchGroupIds, int batchGroupIdOffset)
+    {
+        int[] hashPositions = new int[batchSize];
+        for (int i = 0; i < batchSize; i++) {
+            if (!block.isNull(blockOffset + i)) {
+                hashPositions[i] = getHashPosition(BIGINT.getLong(block, blockOffset + i), mask);
+            }
+        }
+
+        for (int i = 0; i < batchSize; i++) {
+            if (!block.isNull(blockOffset + i)) {
+                batchGroupIds[batchGroupIdOffset + i] = this.groupIds[hashPositions[i]];
+            }
+            else {
+                batchGroupIds[batchGroupIdOffset + i] = NULL_GROUP_ID_MARKER;
+            }
+        }
+
+        for (int i = 0; i < batchSize; i++) {
+            if (batchGroupIds[batchGroupIdOffset + i] >= 0) {
+                long value = BIGINT.getLong(block, blockOffset + i);
+                long storedValue = values[hashPositions[i]];
+                // Same as
+                // if (value != storedValue)
+                //   batchGroupIds[batchGroupIdOffset + i] = -1;
+                // but without explicit branches
+                int match = value == storedValue ? 1 : 0;
+                batchGroupIds[batchGroupIdOffset + i] = (batchGroupIds[batchGroupIdOffset + i] + 1) * match - 1;
+            }
+        }
+
+        if (nullGroupId < 0) { // Assigning new null group id adds another branch inside a loop so we want to do it only once
+            for (int i = 0; i < batchSize; i++) {
+                if (batchGroupIds[batchGroupIdOffset + i] == NULL_GROUP_ID_MARKER) {
+                    if (nullGroupId < 0) {
+                        // set null group id
+                        nullGroupId = nextGroupId++;
+                    }
+                    batchGroupIds[batchGroupIdOffset + i] = nullGroupId;
+                }
+                else if (batchGroupIds[batchGroupIdOffset + i] == -1) {
+                    batchGroupIds[batchGroupIdOffset + i] = putIfAbsent(BIGINT.getLong(block, blockOffset + i), hashPositions[i]);
+                }
+            }
+        }
+        else {
+            for (int i = 0; i < batchSize; i++) {
+                if (batchGroupIds[batchGroupIdOffset + i] == NULL_GROUP_ID_MARKER) {
+                    batchGroupIds[batchGroupIdOffset + i] = nullGroupId;
+                }
+                else if (batchGroupIds[batchGroupIdOffset + i] == -1) {
+                    batchGroupIds[batchGroupIdOffset + i] = putIfAbsent(BIGINT.getLong(block, blockOffset + i), hashPositions[i]);
+                }
+            }
+        }
+    }
+
+    private void batchedPutIfAbsentNoNull(Block block, int blockOffset, int batchSize, long[] batchGroupIds, int batchGroupIdOffset)
+    {
+        int[] hashPositions = new int[batchSize];
```

Comment:
What do you think about reusing this array across multiple `batchedPutIfAbsent` calls to reduce GC pressure?

Reply:
Caching objects in java is usually an anti-pattern. I don't see how should this work.

Reply:
almost all the things we do for performance or efficiency are java "anti-patterns" but I don't think we care about GC that much at the moment so feel free to skip this advice.

Reply:
'almost all the things we do for performance or efficiency are java "anti-patterns"' - that's a good point.
My argument behind this anti-pattern thing is that GC likes a lot of eden collections. So producing a lot of short living objects will trigger more quicker eden collections which is actually good, becaue GC is made to work that way.
Of course this is just a theory. We can easily spawn a three-day discussion about that.

One way or another let;s be honest. The performance difference would be too small to notice. But thanks for bringing that up